src/cwe_checker_lib/src/analysis/stack_alignment_substitution/mod.rs

```diff
@@ -0,0 +1,401 @@
+fn substitute(
+    exp: &mut Expression,
+    alignment: u32,
+    journaled_sp: &mut u32,
+    tid: Tid,
+) -> Vec<LogMessage> {
+    let mut log: Vec<LogMessage> = vec![];
+
+    if let Expression::BinOp { op, lhs: _, rhs } = exp {
+        if let Expression::Const(a) = &**rhs {
+            let bitmask: u32 = ApInt::try_to_u32(&ApInt::into_negate(a.clone())).unwrap();
+
+            let offset = journaled_sp.checked_rem(bitmask).unwrap_or(0);
+
+            match op {
+                BinOpType::IntAnd => {
+                    *op = BinOpType::IntSub;
+                    *journaled_sp -= offset;
+                    if bitmask != alignment {
+                        log.push(LogMessage::new_info("Unexpected alignment").location(tid));
+                    }
+                }
+                BinOpType::IntOr => {
+                    *op = BinOpType::IntAdd;
+                    *journaled_sp += offset;
+                    if bitmask != alignment {
+                        log.push(LogMessage::new_info("Unexpected alignment").location(tid));
+                    }
+                }
+                _ => {
+                    log.push(LogMessage::new_info("Unsubstitutable Operation on SP").location(tid))
+                }
+            };
+            *rhs = Box::new(Expression::Const(ApInt::from_u32(offset)));
+        }
+    } else {
+        log.push(LogMessage::new_info("Unsubstitutable Operation on SP").location(tid))
+    }
+    log
+}
+
+/// Updates current stackpointer value by given Constant.
+fn journal_sp_value(sp: &mut u32, is_plus: bool, val: &Expression) {
+    match val {
+        Expression::Const(con) => {
+            if is_plus {
+                *sp += con.try_to_u32().unwrap()
+            } else {
+                *sp -= con.try_to_u32().unwrap()
+            }
+        }
+        _ => todo!(),
+    }
+}
+
+/// Substitutes logical AND and OR on the stackpointer register by SUB and ADD.
+/// Expressions are changed to use constants w.r.t the provided bit mask.
+pub fn substitute_and_on_stackpointer(project: &mut Project) -> Vec<LogMessage> {
+    // for sanity check
+    let sp_alignment = match project.cpu_architecture.as_str() {
+        "x86_32" => 16,
+        "x86_64" => 16,
+        "arm32" => 4,
+        _ => 0,
+    };
+
+    let journaled_sp: &mut u32 = &mut 10000000; // 128-Byte aligned
+    let mut log: Vec<LogMessage> = vec![];
+
+    for sub in project.program.term.subs.values_mut() {
+        for blk in sub.term.blocks.iter_mut() {
+            for def in blk.term.defs.iter_mut() {
+                if let Def::Assign { var, value } = &mut def.term {
+                    if *var == project.stack_pointer_register {
+                        if let Expression::BinOp { op, lhs, rhs } = value {
+                            if *lhs
+                                == Box::new(Expression::Var(project.stack_pointer_register.clone()))
```

Comment:
It is not guaranteed that the stack pointer is the left operand. It may also be the right operand. Also: We should handle the `else` case here since it would denote a case where something unexpected happens with the stack pointer.